internal/contour/cachehandler.go

```diff
@@ -42,30 +48,28 @@ func (ch *CacheHandler) OnChange(dag *dag.DAG) {
 	timer := prometheus.NewTimer(ch.CacheHandlerOnUpdateSummary)
 	defer timer.ObserveDuration()
 
-	ch.updateSecrets(dag)
-	ch.updateListeners(dag)
-	ch.updateRoutes(dag)
-	ch.updateClusters(dag)
+	ch.SnapshotHandler.UpdateSnapshot(map[xds.ResponseType][]xds.Resource{
+		xds.Cluster:  ch.updateClusters(dag),
+		xds.Route:    ch.updateRoutes(dag),
+		xds.Listener: ch.updateListeners(dag),
+		xds.Secret:   ch.updateSecrets(dag),
+	})
 
 	ch.SetDAGLastRebuilt(time.Now())
 }
 
-func (ch *CacheHandler) updateSecrets(root dag.Visitable) {
-	secrets := visitSecrets(root)
-	ch.SecretCache.Update(secrets)
+func (ch *CacheHandler) updateSecrets(root dag.Visitable) []xds.Resource {
+	return ch.SecretTranslator.Contents(visitSecrets(root))
```

Comment:
`visitSecrets` (and other similar functions) currently returns a map, but that seems unnecessary now. If we changed them to just return a slice, then we could also drop the code in the translators that extract a slice of map values.